src/binary/uint.rs

```diff
@@ -26,23 +30,45 @@ impl DecodedUInt {
             ));
         }
 
-        // Create a stack-allocated buffer to hold the data we're going to read in.
-        let mut buffer = [0u8; MAX_UINT_SIZE_IN_BYTES];
+        if length <= UINT_STACK_BUFFER_SIZE {
+            let buffer = &mut [0u8; UINT_STACK_BUFFER_SIZE];
+            DecodedUInt::read_using_buffer(data_source, length, buffer)
+        } else {
+            // We're reading an enormous int. Heap-allocate a Vec to use as storage.
+            let mut buffer = vec![0u8; length];
+            DecodedUInt::read_using_buffer(data_source, length, buffer.as_mut_slice())
+        }
+    }
```

Comment:
PR tour: when decoding a `UInt`, the reader will use a stack-allocated buffer for values that are `<= 16` bytes. If a particularly large `UInt` is encountered, the reader will allocate an appropriately sized `Vec` to decode it. This should be very rare indeed; however, it does occur in several of the `ion-tests`. The largest integer in `ion-tests` is some 1200+ bytes at the moment. At the top of this method, we enforce an overall maximum allowable size, currently set to 2048 bytes. This is to prevent this from becoming a denial-of-service vector.

Reply:
> I wonder what the largest integer we've encountered or heard of in real data is?

Excellent question. It's not hard to find ion-java client code using `IonReader#bigIntegerValue()`, but it's impossible to know whether their data requires that or it's just defensive programming.

A bit of code archeology: back when it was a hobby project,`ion-rust` originally only supported `i64` representations of integer. When Rust v1.26 was released, it introduced [a stable `i128` data type](https://doc.rust-lang.org/std/primitive.i128.html). I briefly planned to make that the go-to representation for integer; it was stack allocated and had fast intrinsic operations relative to `BigInt`. Ultimately though, I went with `BigInt` because I knew there were [spec conformance tests that featured enormous integers](https://github.com/amzn/ion-tests/blob/master/iontestdata/good/intBigSize1201.10n). 

We could have just said "we don't support arbitrarily sized ints" and added those test files to a skip list, but I concluded that when 99.999% of integer use cases will fit in an `i64`, we don't lose much by using a `BigInt` instead of an `i128` for the remaining .001%.